presto-hive/src/main/java/com/facebook/presto/hive/S3SelectLineRecordReader.java

```diff
@@ -0,0 +1,200 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.amazonaws.services.s3.model.SelectObjectContentRequest;
+import com.facebook.presto.spi.predicate.TupleDomain;
+import com.facebook.presto.spi.type.TypeManager;
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.WaitStrategies;
+import com.google.common.base.Predicates;
+import io.airlift.log.Logger;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapred.RecordReader;
+import org.apache.hadoop.util.LineReader;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.apache.hadoop.hive.serde.serdeConstants.FIELD_DELIM;
+import static org.apache.hadoop.hive.serde.serdeConstants.LINE_DELIM;
+import static org.apache.hadoop.hive.serde.serdeConstants.SERIALIZATION_FORMAT;
+
+public abstract class S3SelectLineRecordReader
+        implements RecordReader<LongWritable, Text>
+{
+    private static final Logger log = Logger.get(S3SelectLineRecordReader.class);
+
+    private static final long MAX_SLEEP_MS = 200;
+    private static final int MAX_RETRY_NUMBER = 5;
+    private final S3SelectClient selectClient;
+    private InputStream selectObjectContent;
+    final String lineDelimiter;
+    private final List<HiveColumnHandle> columns;
+    private final TupleDomain<HiveColumnHandle> effectivePredicate;
+    private final TypeManager typeManager;
+    private int processedRecords;
+    private int recordsFromS3;
+    private long start;
+    private long pos;
+    private long end;
+    private LineReader reader;
+    private boolean isFirstLine;
+
+    S3SelectLineRecordReader(S3SelectClient selectClient, Path path, long start, long length,
+                             Properties schema, List<HiveColumnHandle> columns, TupleDomain<HiveColumnHandle> effectivePredicate, TypeManager typeManager)
+    {
+        this.selectClient = selectClient;
+        this.columns = columns;
+        this.effectivePredicate = effectivePredicate;
+        this.typeManager = typeManager;
+        this.lineDelimiter = getLineDelimiter(schema);
+        this.processedRecords = 0;
+        this.recordsFromS3 = 0;
+        this.start = start;
+        this.pos = this.start;
+        this.end = this.start + length;
+        this.selectClient.initializeRequest(buildSelectObjectRequest(schema, buildQuery(), path));
```

Comment:
This will cause wrong results. Each file can be divided into many splits  but each split will now run the same query on s3 select on full file so it will consider rows multiple times. Need to mark path as non splittable when creating splits in BackgroundHiveSplitLoader.

Reply:
Nice catch. Sure, will add tests for this.

Reply:
Fixed and tested that we don't set splittable parameter to true when optimizer is enabled.